builder.go

```diff
@@ -82,83 +100,39 @@ func (b *Builder) Build() (*BuildMetadata, error) {
 	var labels []Label
 
 	for _, bp := range b.Group.Group {
-		bpInfo, err := bp.Lookup(b.BuildpacksDir)
-		if err != nil {
-			return nil, err
-		}
-		bpDirName := launch.EscapeID(bp.ID)
-		bpLayersDir := filepath.Join(layersDir, bpDirName)
-		bpPlanDir := filepath.Join(planDir, bpDirName)
-		if err := os.MkdirAll(bpLayersDir, 0777); err != nil {
-			return nil, err
-		}
-
-		if err := os.MkdirAll(bpPlanDir, 0777); err != nil {
-			return nil, err
-		}
-		bpPlanPath := filepath.Join(bpPlanDir, "plan.toml")
-
 		foundPlan := plan.find(bp.noAPI().noHomepage())
 		if api.MustParse(bp.API).Equal(api.MustParse("0.2")) {
 			for i := range foundPlan.Entries {
 				foundPlan.Entries[i].convertMetadataToVersion()
 			}
 		}
-		if err := WriteTOML(bpPlanPath, foundPlan); err != nil {
+
+		bpLayersDir, bpPlanPath, err := preparePaths(bp, foundPlan, layersDir, planDir)
+		if err != nil {
 			return nil, err
 		}
 
-		cmd := exec.Command(
-			filepath.Join(bpInfo.Path, "bin", "build"),
-			bpLayersDir,
-			platformDir,
-			bpPlanPath,
-		)
-		cmd.Dir = appDir
-		cmd.Stdout = b.Out
-		cmd.Stderr = b.Err
-
-		if bpInfo.Buildpack.ClearEnv {
-			cmd.Env = b.Env.List()
-		} else {
-			cmd.Env, err = b.Env.WithPlatform(platformDir)
-			if err != nil {
-				return nil, err
-			}
+		if err := b.runBuildCmd(bp, appDir, bpLayersDir, platformDir, bpPlanPath); err != nil {
```

Comment:
Sounds good... while we're at it, what's the system that we want to be using to organize code within a file? Sometimes we have the structs at the top, sometimes structs grouped with methods, etc.

Reply:
On second thought... what would be the advantage of making `BuildpackTOML` the receiver given that we already have `[]Buildpack` in `Builder.Group.Group`?

Reply:
After experimenting with this approach I think I like how it reads... probably it makes sense to unit-test `BuildpackTOML.Build()` as part of this PR, will add in the next round.

Reply:
There are some differences in how `BuildpackTOML.Detect()` behaves vs `BuildpackTOML.Build()` which may or may not be significant - notably the detect config takes `FullEnv` and `ClearEnv` whereas build config has `Env` (this is because of `setupEnv(config.Env, bpLayersDir)`). Also detect passes the output back to the caller whereas build takes a writer.

Reply:
I think the work you have done so far is a major improvement. We work on tackling all of the inconsistencies over time but this feels like a big step in the right direction.

Reply:
> Sounds good... while we're at it, what's the system that we want to be using to organize code within a file? Sometimes we have the structs at the top, sometimes structs grouped with methods, etc.

One of my default opinions is that the definition should be directly above any function with that type as a receiver (if the definition and method are in the same file).
```
type Foo struct {} // struct with methods
func (*Foo) Method1() {}
func (*Foo) Method2() {}
type Bar struct {} // struct with methods
func (*Bar) Method1() {}
```
stdlib example: https://golang.org/src/net/iprawsock.go#L99

Not sure there is one right answer to this question. If there is one right answer, I don't know it. Whatever feels organized and easy to navigate? In the future I think a new `buildpack` package might help us organize some of our existing types. In this top-level package a lot of things end up in `build.go` that we would probably organize in a larger number of smaller files if they lived in a separate package.